### ISSUE TITLE
Feature request - Ability to set tag element on component without using a prop

### ISSUE THREAD

**Issue by Dobby89** [facebook/react#14368]
I think it would be useful to have the ability to override the containing element tag of a component.

Currently the only way to override the containing tag element type is to set it as a prop when calling the component. For example,

```
const MyButtonComponent = props => <props.tag>{ props.children }</props.tag>

<MyButtonComponent tag="button">I'm a button</MyButtonComponent>

<MyButtonComponent tag="a">I'm an anchor link</MyButtonComponent>
```

It works, but I feel like if we can infer the props.children from whatever is inbetween the component tags, why not provide some more syntactic sugar for setting the tag element as well?

I'm thinking something like `<Component::tag>Content</Component::tag>` . So the previous example would then be rewritten as follows.

```
<MyButtonComponent::button>I'm a button</MyButtonComponent::button>

<MyButtonComponent::a>I'm an anchor link</MyButtonComponent::a>
```

If this is already possible using alternative methods I'll be happy to know about them. I'm sure I've neglected some sort of edge case which might cause issues.

**Comment by Aliin** [comment]
I disagree with the assessment that passing tag type by prop is not good. Props are not mainly used for setting data inside a tag - they are used for all kinds of things, like passing callback functions. They are basically parameters for constructing the react component, which is exactly where something like tag type should go.

I think the best way to deal with this is by using a render prop that contains the tag type you want.

**Comment by Dobby89** [comment]
@Aliin Thanks for your input.

> I disagree with the assessment that passing tag type by prop is not good.

I didn't suggest it was not good. I feel that from a convenience and code readability point of view, this:

```
<MyButtonComponent::button>I'm a button</MyButtonComponent::button>
```

is more readable at a glance than this:

```
<MyButtonComponent tag="button">I'm a button</MyButtonComponent>
```

Don't forget that tag may be one of many props on the component, so in reality the tag prop might disappear amongst all the other props.
